source/distributed/tria_base.cc

```diff
@@ -362,36 +362,125 @@ namespace parallel
+                    }
+                    {
+                      auto it = vertex_to_coinciding_vertex.find(v_n);
+                      if (it != vertex_to_coinciding_vertex.end())
+                        temp = std::min(temp, it->second);
+                    }
+                    vertex_to_coinciding_vertex[v_t] = temp;
+                    vertex_to_coinciding_vertex[v_n] = temp;
+                  }
+              }
+
+      // 1b) compress map: let vertices point to the coinciding vertex with
+      //     the smallest index
+      for (auto &p : vertex_to_coinciding_vertex)
+        {
+          if (p.first == p.second)
+            continue;
+          unsigned int temp = p.second;
+          while (temp != vertex_to_coinciding_vertex[temp])
+            temp = vertex_to_coinciding_vertex[temp];
+          p.second = temp;
+        }
 
+#ifdef DEBUG
+      // check if map is actually compressed
+      for (auto p : vertex_to_coinciding_vertex)
+        {
+          if (p.first == p.second)
+            continue;
+          auto pp = vertex_to_coinciding_vertex.find(p.second);
+          if (pp->first == pp->second)
+            continue;
+          AssertThrow(false, ExcMessage("Map has to be compressed!"));
+        }
+#endif
 
-    std::vector<bool> vertex_of_own_cell(this->n_vertices(), false);
+      // 1c) create a map: smallest index of coinciding index -> all
+      // coinciding indices
+      std::map<unsigned int, std::set<unsigned int>>
+        smallest_coinciding_vertex_to_coinciding_vertices;
+      for (auto p : vertex_to_coinciding_vertex)
+        smallest_coinciding_vertex_to_coinciding_vertices[p.second] =
+          std::set<unsigned int>();
+
+      for (auto p : vertex_to_coinciding_vertex)
+        smallest_coinciding_vertex_to_coinciding_vertices[p.second].insert(
+          p.first);
+
+      // 1d) create a map: vertex -> all coinciding indices
+      for (auto &s : smallest_coinciding_vertex_to_coinciding_vertices)
+        for (auto &ss : s.second)
+          vertex_to_coinciding_vertices[ss] = s.second;
+    }
 
+    // 2) collect vertices belonging to local cells
+    std::vector<bool> vertex_of_own_cell(this->n_vertices(), false);
     for (const auto &cell : this->active_cell_iterators())
       if (cell->is_locally_owned())
-        for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell; ++v)
+        for (auto v = 0u; v < GeometryInfo<dim>::vertices_per_cell; ++v)
           vertex_of_own_cell[cell->vertex_index(v)] = true;
 
+    // 3) for for each vertex belonging to a locally owned cell all ghost
+    //    neighbors (including the periodic own)
     std::map<unsigned int, std::set<dealii::types::subdomain_id>> result;
+
+    // loop over all active ghost cells
     for (const auto &cell : this->active_cell_iterators())
       if (cell->is_ghost())
         {
-          const types::subdomain_id owner = cell->subdomain_id();
+          const auto owner = cell->subdomain_id();
```

Comment:
Also here, I prefer to keep the type - right now, the type is explicit in line 453 via the `std::set` but implicit here. Since this is so little to type, the educational aspect of listing it here is more important than saving 15 characters in my opinion, at least for existing code. (As opposed to say an iterator that comes out of `find` in line 471 below.)